queue.go

```diff
@@ -245,22 +247,53 @@ func (queue *Queue) HardReset(s *Server) {
 	// - task.Delete() writes to a buffered `cancel` channel
 	// - task.Schedule() reads from that buffered channel in a separate goroutine
 	// - When that goroutine sees the task is cancelled, it sets the task value to nil in the tasks map
+	// - Additionally, if a task has already been dispatched then task.Delete() has no effect until after the current
+	//   execution, which depends entirely on the response time of the task's target.
 	//
 	// We need to be certain that we only remove the task from map *after* that completes, otherwise the task name will
-	// be reinserted with the nil value. At the moment the only easy way I can think of is to sleep for a very short
-	// period to allow the tasks' internal goroutines to fire first.
-	time.Sleep(10 * time.Millisecond)
+	// be reinserted with the nil value.
+	isReadyChannel := make(chan bool, 1)
+	tryDeleteTasks := func() {
+		queue.tsMux.Lock()
+		defer queue.tsMux.Unlock()
 
-	queue.tsMux.Lock()
-	defer queue.tsMux.Unlock()
-	for taskName, task := range queue.ts {
-		if task != nil {
-			// The naive "sleep till it deletes" approach described above is too naive...
-			panic("Expected task to be deleted by now!")
+		hasAnyPending := false
+		for taskName, task := range queue.ts {
+			if task == nil {
+				// Task has already been deleted / ran to completion - safe to remove
+				delete(queue.ts, taskName)
+				s.hardDeleteTask(taskName)
+			} else {
+				// Task is still running (or the `cancel` channel has not fired) - will need to wait and retry
+				hasAnyPending = true
+			}
 		}
+		isReadyChannel <- !hasAnyPending
+	}
 
-		delete(queue.ts, taskName)
-		s.hardDeleteTask(taskName)
+	// The timeout applies across all iterations of the for loop.
+	// It is intentionally relatively short, because the internal retry interval is rapid.
+	// If calling code expects some task requests to last longer, it should handle the DEADLINE_EXCEEDED error and retry
+	// on a schedule to suit the application.
+	timeout := time.After(3 * time.Second)
+
+	go tryDeleteTasks()
```

Comment:
Why is there a need for a go routines here? Since there is no asynchronous operation, and the intent is for the caller to wait anyway, I believe a simple loop with a `time.Since(time.Now()) >= 3*time.Second` can suffice.

Reply:
Probably just my inexperience with go threading/async. I suspect you're right, I'll give it a go with a simpler loop.